src/lib.rs

```diff
@@ -740,4 +722,136 @@ impl DependencyNames {
+		}
+	}
+}
+
+impl IndexMut<dependency::Kind> for DependencyNames {
+	fn index_mut(&mut self, index: dependency::Kind) -> &mut DependencyNamesValue {
+		match index {
+			dependency::Kind::Normal => &mut self.normal,
+			dependency::Kind::Development => &mut self.development,
+			dependency::Kind::Build => &mut self.build,
+		}
+	}
+}
+
+#[derive(Debug, Default)]
+struct DependencyNamesValue {
+	by_extern_crate_name :HashMap<String, InternedString>,
+	by_lib_true_snakecased_name :HashMap<String, HashSet<InternedString>>,
+	non_lib :HashSet<InternedString>,
+}
+
+#[derive(Default, Debug, Serialize)]
+struct Outcome {
+	success: bool,
+	unused_deps: BTreeMap<PackageId, OutcomeUnusedDeps>,
+	warning: Option<String>,
+}
+
+impl Outcome {
+	fn print(&self, output: OutputKind, stdout: impl Write) -> io::Result<()> {
+		match output {
+			OutputKind::Human => self.print_human(stdout),
+			OutputKind::Json => self.print_json(stdout),
+		}
+	}
+
+	fn print_human(&self, mut stdout: impl Write) -> io::Result<()> {
+		if self.success {
+			writeln!(stdout, "All deps seem to have been used.")?;
+		} else {
+			writeln!(stdout, "unused dependencies:")?;
+
+			for (member, OutcomeUnusedDeps { normal, development, build, .. }) in &self.unused_deps {
+				fn edge_and_joint(p: bool) -> (char, char) {
+					if p {
+						(' ', '└')
+					} else {
+						('│', '├')
+					}
+				}
+
+				writeln!(stdout, "`{}`", member)?;
+
+				for (deps, (edge, joint), prefix) in &[
+					(normal, edge_and_joint(development.is_empty() && build.is_empty()), ""),
+					(development, edge_and_joint(build.is_empty()), "dev-"),
+					(build, (' ', '└'), "build-"),
+				] {
+					if !deps.is_empty() {
+						writeln!(stdout, "{}─── {}dependencies", joint, prefix)?;
+						let mut deps = deps.iter().peekable();
+						while let Some(dep) = deps.next() {
+							let joint = if deps.peek().is_some() {
+								'├'
+							} else {
+								'└'
+							};
+							writeln!(stdout, "{}    {}─── {:?}", edge, joint, dep)?;
+						}
+					}
+				}
+			}
+
+			if let Some(warning) = &self.warning {
+				write!(stdout, "{}", warning)?;
+			}
+		}
+		stdout.flush()
+	}
+
+	fn print_json(&self, mut stdout: impl Write) -> io::Result<()> {
+		let json = serde_json::to_string(self).expect("should not fail");
+		writeln!(stdout, "{}", json)?;
+		stdout.flush()
+	}
+}
+
+#[derive(Debug, Serialize)]
+struct OutcomeUnusedDeps {
+	manifest_path: String,
+	normal: BTreeSet<InternedString>,
+	development: BTreeSet<InternedString>,
+	build: BTreeSet<InternedString>,
+}
+
+impl OutcomeUnusedDeps {
+	fn new(manifest_path: &Path) -> CargoResult<Self> {
+		let manifest_path = manifest_path
+			.to_str()
+			.ok_or_else(|| failure::format_err!("{:?} is not valid utf-8", manifest_path))?
+			.to_owned();
+
+		Ok(Self {
+			manifest_path,
+			normal: BTreeSet::new(),
+			development: BTreeSet::new(),
+			build: BTreeSet::new(),
+		})
+	}
+}
+
+#[derive(EnumString, EnumVariantNames, Clone, Copy, Debug)]
```

Comment:
Do we really need a separate crate for this? I think it would be better if you made the code manually, if that's possible.

Reply:
It is actually possible. `strum` saves only 14 lines.

```diff
-#[derive(EnumString, EnumVariantNames, Clone, Copy, Debug)]
-#[strum(serialize_all = "kebab-case")]
+use std::str::FromStr;
+
+#[derive(Clone, Copy, Debug)]
 enum OutputKind {
        Human,
        Json,
 }
+
+impl OutputKind {
+       const VARIANTS: &'static [&'static str] = &["human", "json"];
+}
+
+impl FromStr for OutputKind {
+       type Err = &'static str;
+
+       fn from_str(s: &str) -> std::result::Result<Self, &'static str> {
+               match s {
+                       "human" => Ok(Self::Human),
+                       "json" => Ok(Self::Json),
+                       _ => Err(r#"expected "human" or "json""#),
+               }
+       }
+}
```

But I don't think it causes bloating so much. And `strum` is a widely used reliable crate. It has 1M downloads and [88 reverse dependencies](https://crates.io/crates/strum/reverse_dependencies).

![graph](https://user-images.githubusercontent.com/14125495/73593708-4bcf6180-454a-11ea-8ac6-e880e8e97a91.png)

```console
$ pwd
/home/ryo/src/github.com/Peternator7/strum
$ tokei
-------------------------------------------------------------------------------
 Language            Files        Lines         Code     Comments       Blanks
-------------------------------------------------------------------------------
 Markdown                2          157          157            0            0
 Rust                   35         2723         2135          243          345
 Shell                   1           15           12            0            3
 TOML                    5          103           88            0           15
 YAML                    1          129           47           66           16
-------------------------------------------------------------------------------
 Total                  44         3127         2439          309          379
-------------------------------------------------------------------------------
$ touch ./strum/src/lib.rs ./strum_macros/src/**/*.rs && cargo build
   Compiling strum_macros v0.17.1 (/home/ryo/src/github.com/Peternator7/strum/strum_macros)
   Compiling strum v0.17.1 (/home/ryo/src/github.com/Peternator7/strum/strum)
   Compiling strum_tests v0.15.0 (/home/ryo/src/github.com/Peternator7/strum/strum_tests)
    Finished dev [unoptimized + debuginfo] target(s) in 1.28s
$ touch ./strum/src/lib.rs ./strum_macros/src/**/*.rs && cargo build --release
   Compiling strum_macros v0.17.1 (/home/ryo/src/github.com/Peternator7/strum/strum_macros)
   Compiling strum v0.17.1 (/home/ryo/src/github.com/Peternator7/strum/strum)
   Compiling strum_tests v0.15.0 (/home/ryo/src/github.com/Peternator7/strum/strum_tests)
    Finished release [optimized] target(s) in 6.19s
```

Reply:
I don't think that two more dependent crates with additional 2000 lines of code are worth the 14 (pretty straightforward) lines. I'd only use it if we then make heavier use of it.

Reply:
Also btw, your compile time tests also compile the `strum_tests` crate. It means 3.18s if you do `cargo build --release -p strum -p strum_macro` instead of `cargo build --release`.

Reply:
> I'd only use it if we then make heavier use of it.

I understand. Removed the `strum` dependency and `cargo update` part.